### PR TITLE
JAMES-3460 MailboxChangeListener implementation

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      case mailboxACLUpdated: MailboxACLUpdated => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(mailboxACLUpdated.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(mailboxACLUpdated.getMailboxId).asJava, List().asJava))
+      case mailboxDeletion: MailboxDeletion => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(mailboxDeletion.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List().asJava, List(mailboxDeletion.getMailboxId).asJava))
+      case messageAdded: Added => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(messageAdded.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(messageAdded.getMailboxId).asJava, List().asJava))
+      case messageDeleted: Expunged => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(messageDeleted.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(messageDeleted.getMailboxId).asJava, List().asJava))
+      case _ => SMono.empty
+    }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala:50]: Implement isHandling... 

This optimizes the event delivery and publish the event only if we are really ahndling it.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      case mailboxACLUpdated: MailboxACLUpdated => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(mailboxACLUpdated.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(mailboxACLUpdated.getMailboxId).asJava, List().asJava))
+      case mailboxDeletion: MailboxDeletion => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(mailboxDeletion.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List().asJava, List(mailboxDeletion.getMailboxId).asJava))
+      case messageAdded: Added => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(messageAdded.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(messageAdded.getMailboxId).asJava, List().asJava))
+      case messageDeleted: Expunged => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(messageDeleted.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(messageDeleted.getMailboxId).asJava, List().asJava))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala:47]: `ZonedDateTime.now()` => use a clock.

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala:47]: Don't repeat yourself:

I think you need a factory method (event => MailboxChange) and do the .save only once.

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/change/MailboxChangeListenerTest.scala [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    val state = State.of(UUID.randomUUID)
+    repository.save(MailboxChange.of(ACCOUNT_ID, state, ZonedDateTime.now, List[MailboxId](TestId.of(0)).asJava, List().asJava, List().asJava)).block()
+
+    mailboxManager.deleteMailbox(inboxId, mailboxSession)
+
+    assertThat(repository.getSinceState(ACCOUNT_ID, state, None.toJava).block().getDestroyed)
+      .containsExactly(inboxId)
+  }
+}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/change/MailboxChangeListenerTest.scala:159]: You did not test flag updates.

**Review by chibenwa — COMMENTED** [review]
Looks like good work to me ;-)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/MailboxChange.java [hunk 4]
```diff
@@ -82,6 +94,40 @@ public static MailboxChange of(AccountId accountId, State state,  ZonedDateTime
...
+            MailboxDeletion mailboxDeletion = (MailboxDeletion) event;
+            return Optional.of(MailboxChange.of(AccountId.fromUsername(mailboxDeletion.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(), ImmutableList.of(), ImmutableList.of(mailboxDeletion.getMailboxId())));
+        }
+        if (event instanceof Added) {
+            Added messageAdded = (Added) event;
+            return Optional.of(MailboxChange.of(AccountId.fromUsername(messageAdded.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(), ImmutableList.of(messageAdded.getMailboxId()), ImmutableList.of()));
+        }
+        if (event instanceof FlagsUpdated) {
+            FlagsUpdated messageAdded = (FlagsUpdated) event;
+            return Optional.of(MailboxChange.of(AccountId.fromUsername(messageAdded.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(), ImmutableList.of(messageAdded.getMailboxId()), ImmutableList.of()));
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/MailboxChange.java:121]: We are interested in this only if the SEEN flag is updated.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala [hunk 5]
```diff
@@ -19,34 +19,23 @@
...
-      case messageDeleted: Expunged => mailboxChangeRepository.save(MailboxChange.of(AccountId.fromUsername(messageDeleted.getUsername), State.of(UUID.randomUUID()), ZonedDateTime.now(), List().asJava, List(messageDeleted.getMailboxId).asJava, List().asJava))
-      case _ => SMono.empty
-    }
+    MailboxChange.fromEvent(event)
+      .map(mailboxChangeRepository.save(_))
+      .orElse(SMono.empty[Void].asJava)
 
   override def getDefaultGroup: Group = MailboxChangeListenerGroup()
+
+  override def isHandling(event: Event): Boolean = event.isInstanceOf[MailboxEvent]
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/MailboxChangeListener.scala:40]: How about `MailboxChange.fromEvent(event).isDefined` ?

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/change/MailboxChangeListenerTest.scala [hunk 6]
```diff
@@ -126,6 +127,23 @@ class MailboxChangeListenerTest {
...
+
+    val state = State.of(UUID.randomUUID)
+    repository.save(MailboxChange.of(ACCOUNT_ID, state, ZonedDateTime.now, List[MailboxId](TestId.of(0)).asJava, List().asJava, List().asJava)).block()
+
+    messageManager.setFlags(new Flags(Flags.Flag.SEEN), FlagsUpdateMode.ADD, MessageRange.all(), mailboxSession)
+
+    assertThat(repository.getSinceState(ACCOUNT_ID, state, None.toJava).block().getUpdated)
+      .containsExactly(inboxId)
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/change/MailboxChangeListenerTest.scala:146]: Add a test with a flag update test without the SEEN flag, it should not update things!

**Comment by chibenwa** [comment]
```
[607469e5041658d555a5f7f24b6aab07f1b53453] [ERROR]   MemorySpamAssassinContractTest.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations » ConditionTimeout
```

test this please

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**